lib/active_merchant/billing/gateways/payu_latam.rb

```diff
@@ -183,15 +183,25 @@ def billing_address_fields(options)
         billing_address
       end
 
-      def add_buyer(post, options)
+      def add_buyer(post, payment_method, options)
         buyer = {}
-        buyer[:fullName] = options[:buyer_name] if options[:buyer_name]
-        buyer[:dniNumber] = options[:buyer_dni_number] if options[:buyer_dni_number]
-        buyer[:dniType] = options[:buyer_dni_type] if options[:buyer_dni_type]
-        buyer[:cnpj] = options[:buyer_cnpj] if options[:buyer_cnpj] && options[:payment_country] == 'BR'
-        buyer[:emailAddress] = options[:buyer_email] if options[:buyer_email]
-        buyer[:contactPhone] = options[:shipping_address][:phone] if options[:shipping_address]
-        buyer[:shippingAddress] = shipping_address_fields(options) if options[:shipping_address]
+        if buyer_hash = options[:buyer]
+          buyer[:fullName] = buyer_hash[:name]
+          buyer[:dniNumber] = buyer_hash[:dni_number]
+          buyer[:dniType] = buyer_hash[:dni_type]
+          buyer[:cnpj] = buyer_hash[:cnpj] if options[:payment_country] == 'BR'
+          buyer[:emailAddress] = buyer_hash[:email]
+          buyer[:contactPhone] = options[:shipping_address][:phone] if options[:shipping_address]
+          buyer[:shippingAddress] = shipping_address_fields(options) if options[:shipping_address]
+        else
+          buyer[:fullName] = payment_method.name.strip
+          buyer[:dniNumber] = options[:dni_number]
+          buyer[:dniType] = options[:dni_type]
+          buyer[:cnpj] = options[:cnpj] if options[:payment_country] == 'BR'
+          buyer[:emailAddress] = options[:email]
+          buyer[:contactPhone] = (options[:shipping_address][:phone] if options[:shipping_address]) || (options[:billing_address][:phone] if options[:billing_address])
```

Comment:
Seems like we check 2 places here for a phone number but on line 194 only under the `shipping_address`. Wondering if there's a reason for the discrepancy?

Reply:
So, a shipping address may be available independent of specified buyer data, so even if the payer's info is substituted in for the buyer fields, they may have specified a different contact phone number for shipping. If there is buyer data provided, the contact phone number should only be picked from the buyer's shipping address. Does that make sense?

Reply:
I think so. I also think I may not have full context given the purpose of these fields and how to properly formulate them.